### ISSUE TITLE
Spam log - unable to remove filesystem

### ISSUE THREAD

**Issue by typia** [rancher/rancher#7016]
**Rancher Version:**
1.2
**Docker Version:**
1.12.3
**OS and where are the hosts located? (cloud, bare metal, etc):**
Debian 8
**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single Node rancher external DB
**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle
**Steps to Reproduce:**
Only add nfs-rancher stack
**Results:** 

spam log 

2016-12-12 13:44:20,946 ERROR [67d957a6-36f2-480e-a6cc-606dd81fe718:9476] [volumeStoragePoolMap:236] [volumestoragepoolmap.remove] [] [cutorService-13] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm:
device or resource busy
2016-12-12 13:44:20,958 ERROR [:] [] [] [] [cutorService-16] [.e.s.i.ProcessInstanceDispatcherImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for e15f4408443ea1065475127023d95526f45ec23aa9e4e13536fb90ce3564dc43: remove /var/lib/docker/containers/e15f4408443ea1065475127023d95526f45ec23aa9e4e13536fb90ce3564dc43/shm: device or resource busy
2016-12-12 13:44:20,990 ERROR [:] [] [] [] [cutorService-13] [.e.s.i.ProcessInstanceDispatcherImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm: device or resource busy
2016-12-12 13:44:35,920 ERROR [de5f5bd7-7171-4687-abbd-4117991d93e1:9476] [volumeStoragePoolMap:236] [volumestoragepoolmap.remove] [] [cutorService-15] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm:
device or resource busy
2016-12-12 13:44:35,946 ERROR [:] [] [] [] [cutorService-15] [.e.s.i.ProcessInstanceDispatcherImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm: device or resource busy
2016-12-12 13:44:50,921 ERROR [0dc500c8-d60b-446e-b576-72440cfed6af:9476] [volumeStoragePoolMap:236] [volumestoragepoolmap.remove] [] [ecutorService-9] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm:
device or resource busy
2016-12-12 13:44:50,950 ERROR [:] [] [] [] [ecutorService-9] [.e.s.i.ProcessInstanceDispatcherImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm: device or resource busy
2016-12-12 13:45:05,940 ERROR [05aaccbd-a812-4510-bc19-944069720065:9476] [volumeStoragePoolMap:236] [volumestoragepoolmap.remove] [] [ecutorService-5] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm:
device or resource busy
2016-12-12 13:45:05,966 ERROR [:] [] [] [] [ecutorService-5] [.e.s.i.ProcessInstanceDispatcherImpl] Agent error for [storage.volume.remove.reply;agent=46]: Error response from daemon: Unable to remove filesystem for bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d: remove /var/lib/docker/containers/bfe3852bccc63f18cf71d4ee38d4de2e4de5070ab454586d53356d69e275f02d/shm: device or resource busy

**Expected:**

No spam log :)

**Comment by stefanlasiewski** [comment]
I also had several thousand of these errors in my logs. I was running Docker 1.12.2 and 1.12.3.

I updated to Docker 1.12.4, and the messages went away as soon as Docker restarted with the new version.

There are many potential causes of this error message, but I wonder if it was fixed by Docker #29083, listed at https://github.com/docker/docker/releases/tag/v1.12.4

* Fix issue where volume metadata was not removed #[29083](https://github.com/docker/docker/pull/29083)

**Comment by Sturgelose** [comment]
I have the same problem since updating to 1.2.0 and also in 1.2.1. I tried disabling the CAdvisor container I had thinking it might be the problem but still it gets stuck. If I restart the docker service, the dead containers disappear, though....

I recently updated to docker 1.12.5 for Debian and still the same :(

Also, tried to find which process is keeping the resource busy with `lsof` but there is nothing to be seen, so no idea what is happening....

Running in:
```
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 5.883 GiB
```

**Comment by pulberg** [comment]
Same issue here, RHEL7.2, Rancher v1.2.1, upgraded Docker 1.12.4 --> 1.12.5, no change

Full error output - 
https://gist.github.com/pulberg/5ffe71cba5ab0bb52bd808df3d909890

**Comment by cllunsford** [comment]
Seeing the same issue after upgrading to Rancher v1.2.1.  Docker v1.11.2, Ubuntu 14.04

**Comment by cllunsford** [comment]
This looks to be very similar to #6639 .  It appears that the rancher-agent container is holding onto a mount for `/var/lib/docker/containers/[container-id]/shm` preventing the deletion.

docker daemon log:
```
time="2016-12-19T14:42:28.477546954Z" level=error msg="Handler for DELETE /v1.22/containers/6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854 returned error: Unable to remove filesystem for 6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854: remove /var/lib/docker/containers/6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854/shm: device or resource busy" 
```

Finding process with mount:
```grep "6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854" /proc/*/mountinfo 2>/dev/null | awk -F'/' '{print $3}' | xargs ps -p
  PID TTY      STAT   TIME COMMAND
20033 ?        Ss     0:00 /bin/bash /run.sh run
27592 ?        Sl     0:06 /var/lib/cattle/pyagent/agent
27828 ?        Sl     0:04 host-api -cadvisor-url http://127.0.0.1:9344 -logtostderr=true -ip 0.0.0.0 -po
```

If I exec into the rancher-agent, I can confirm that the mount exists:
``` sudo docker exec -it 803c9c18cc5e /bin/bash
$ mount | grep "6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854"                                
shm on /var/lib/docker/containers/6f55ec60dc88cc31a03edb256a2e4165168e8aa5afd0a07f93d0c74feef7d854/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
```

Restarting the `rancher/agent` and `rancher/network-manager` clears the dead containers.

**Comment by ekristen** [comment]
I'm seeing this exact same issue. Please advise.

**Comment by stefanlasiewski** [comment]
Well, turns out that I'm still having this problem. It just took a couple days to re-appear.

Rancher 1.2.0, CentOS 78.2, Docke 1.12.4, Cattle, Devicemapper with direct-lvm.

Are these errors perhaps from containers which are bind mounting /var/lib/docker/ ? Does this Troubleshooting step offer a fix?

https://docs.docker.com/engine/admin/troubleshooting_volume_errors/#/error-unable-to-remove-filesystem

**Comment by vincent99** [comment]
This is a [docker bug](https://github.com/docker/docker/issues/20560) that cAdvisor is good at triggering;  I thought it was fixed in newer 1.12.x-es.  But Rancher 1.2 does not run cAdvisor anymore, so it's unclear why you'd still have it running..

**Comment by ekristen** [comment]
Rancher 1.2.1, Docker 1.12.1, Ubuntu 14.04.5, Cattle, AUFS -- This is a "FRESH" install of rancher on top of some existing servers, I never once had a problem until after installing rancher, then started getting these errors :/

**Comment by cllunsford** [comment]
@vincent99 In the issue you linked, the problem appears to be that cadvisor bind mounts `/var/lib/docker`. (https://github.com/docker/docker/issues/20560#issuecomment-187157369)  Both the `rancher/agent` and `rancher/network-manager` are doing the same thing:

`rancher/agent`
```
"Mounts": [
            {
                "Source": "/var/lib/docker",
                "Destination": "/var/lib/docker",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
...
```

`rancher/network-manager`:
```
        "Mounts": [
            {
                "Source": "/var/lib/docker",
                "Destination": "/var/lib/docker",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            },
...
```

**Comment by deniseschannon** [comment]
Please reference this bug for the fix. We are releasing a 1.2.x for a fix: https://github.com/rancher/rancher/issues/7152

**Comment by deniseschannon** [comment]
As a workaround for now you can try restarting your network manager to clear up the errors about being unable to remove filesystems until the proper fix is released with v1.2.2.

**Comment by deniseschannon** [comment]
As #7152 is closed, when 1.2.2 is released, you will be able to upgrade and have these issues fixed.
